### ISSUE TITLE
[IMPORTANT] - InvalidApiVersionParameter in VMSS

### ISSUE THREAD

**Issue by marcosnils** [Azure/azure-sdk-for-go#805]
Hey there, just updated my SDK to the latest 11.1.0-beta version and I'm getting this error when trying to list the instances in a VMSS using the public azure cloud.
```
2017/10/10 17:06:30 compute.VirtualMachineScaleSetVMsClient#List: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidApiVersionParameter" Messa
ge="The api-version '2017-03-30' is invalid. The supported versions are '2017-08-01,2017-06-01,2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,20
13-03-01,2014-02-26,2014-04'."
```

What's even more strange is that I've tried downgrading my SDK to the  previous version (10.3.1-beta) and still doesn't work.

```
2017/10/10 17:00:07 compute.VirtualMachineScaleSetVMsClient#List: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidApiVersionParameter" Messa
ge="The api-version '2016-04-30-preview' is invalid. The supported versions are '2017-08-01,2017-06-01,2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-
01-01,2013-03-01,2014-02-26,2014-04'."
```

Seems to me like something is really broken here, either from the SDK versioning POV or the Azure API versioning. Now I'm stuck in this situation and I can't get my code to work even if I try to rollback. Any suggestions on how to proceed?

**Comment by marcosnils** [comment]
How come the API Version is hardcoded in the code and we can't change it?

**Comment by marcosnils** [comment]
Managed to find the issue and it was in my side, my code wasn't populating the vmss name and the resource group correctly and those were being set as empty in the url. I realized after enabling logging in the SDK and checking the output URL

```
pwd        | test2017/10/10 19:22:30 Sending GET https://management.azure.com/subscriptions/027a4841-76cc-4b35-8361-b4986a761635/resourceGroups//providers/Microsoft.Compute/virtualMachineScaleSets//virtualMachines?api-version=2017-03-30
pwd        | test2017/10/10 19:22:31 GET https://management.azure.com/subscriptions/027a4841-76cc-4b35-8361-b4986a761635/resourceGroups//providers/Microsoft.Compute/virtualMachineScaleSets//virtualMachines?api-version=2017-03-30 received 400 Bad Request
pwd        | 2017/10/10 19:22:31 compute.VirtualMachineScaleSetVMsClient#List: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidApiVersionParameter" Message="The api-version '2017-03-30' is invalid. The supported versions are '2017-08-01,2017-06-01,2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'."
```
I believe that the Azure SDK should return a more representative error instead of api version error.

**Comment by marstr** [comment]
Howdy @marcosnils, sorry to hear you're running into trouble! We've had a few other reports of this, and I'm currently in the middle of investigating the situation. The particular issue you're hitting smells like an [API Specs](https://github.com/Azure/azure-rest-api-specs) or service issue to me. **I'm going to close this issue as a duplicate of #803**, but feel free to continue the conversation here, and definitely alert us if you continue to be blocked once #803 is resolved.

> @marcosnils says:
> How come the API Version is hardcoded in the code and we can't change it?

There are a few reasons why we hard-code the API Version in the preparers, but important part is that we do this to add guard rails to protect people from inadvertently sending requests that are bound to fail because something changed between API Versions. To unblock you in the short-term, I should mention you _can_ bypass the hardcoded version by writing your own `Preparer` that wraps the generated `Preparer` but updates the Request's API-Version header.
Long-term, what we're doing is generating out code for each API Version. You can see an experimental branch with what the SDK would like in that world here: [azure-sdk-for-go/services @ experimental/allAPIVersions](https://github.com/Azure/azure-sdk-for-go/tree/2fd1c7e96976e2a98c9e51cca51ba1aab124c3d4/services). In this world, you both get guard-rails and the ability to easily change the targeted API Version. It draws attention to the real problem with today's SDK which is that it can only hold models for a single API Versions.

**Comment by marcosnils** [comment]
Thx for the update and the detailed information about the SDK desgin upcoming changes. As you might have seen already, i was hitting a very dumb issue from my side, but the Azure response was totally confusing me
